File: src/network/libqauthenticator.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const std = @import("std");
pub const map_constu8_qtcqvariant = std.StringHashMapUnmanaged(QtC.QVariant);

/// https://doc.qt.io/qt-6/qauthenticator.html
pub const qauthenticator = struct {
    /// New constructs a new QAuthenticator object.
    ///
    ///
    pub fn New() QtC.QAuthenticator {
        return qtc.QAuthenticator_new();
    }

    /// New2 constructs a new QAuthenticator object.
    ///
    /// ``` other: QtC.QAuthenticator ```
    pub fn New2(other: ?*anyopaque) QtC.QAuthenticator {
        return qtc.QAuthenticator_new2(@ptrCast(other));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qauthenticator.html#operator=)
    ///
    /// ``` self: QtC.QAuthenticator, other: QtC.QAuthenticator ```
    pub fn OperatorAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QAuthenticator_OperatorAssign(@ptrCast(self), @ptrCast(other));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qauthenticator.html#operator==)
    ///
    /// ``` self: QtC.QAuthenticator, other: QtC.QAuthenticator ```
    pub fn OperatorEqual(self: ?*anyopaque, other: ?*anyopaque) bool {
        return qtc.QAuthenticator_OperatorEqual(@ptrCast(self), @ptrCast(other));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qauthenticator.html#operator!=)
    ///
    /// ``` self: QtC.QAuthenticator, other: QtC.QAuthenticator ```
    pub fn OperatorNotEqual(self: ?*anyopaque, other: ?*anyopaque) bool {
        return qtc.QAuthenticator_OperatorNotEqual(@ptrCast(self), @ptrCast(other));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qauthenticator.html#user)
    ///
    /// ``` self: QtC.QAuthenticator, allocator: std.mem.Allocator ```
    pub fn User(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QAuthenticator_User(@ptrCast(self));
        defer qtc.libqt_string_free(@constCast(&_str));
        const _ret = allocator.alloc(u8, _str.len) catch @panic("qauthenticator.User: Memory allocation failed");
        for (0.._str.len) |_i| {
            _ret[_i] = _str.data[_i];
        }
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qauthenticator.html#setUser)
    ///
    /// ``` self: QtC.QAuthenticator, user: []const u8 ```
    pub fn SetUser(self: ?*anyopaque, user: []const u8) void {
        const user_str = qtc.struct_libqt_string{
            .len = user.len,
            .data = @constCast(user.ptr),
        };
        qtc.QAuthenticator_SetUser(@ptrCast(self), user_str);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qauthenticator.html#password)
    ///
    /// ``` self: QtC.QAuthenticator, allocator: std.mem.Allocator ```
    pub fn Password(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QAuthenticator_Password(@ptrCast(self));
        defer qtc.libqt_string_free(@constCast(&_str));
        const _ret = allocator.alloc(u8, _str.len) catch @panic("qauthenticator.Password: Memory allocation failed");
        for (0.._str.len) |_i| {
            _ret[_i] = _str.data[_i];
        }
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qauthenticator.html#setPassword)
    ///
    /// ``` self: QtC.QAuthenticator, password: []const u8 ```
    pub fn SetPassword(self: ?*anyopaque, password: []const u8) void {
        const password_str = qtc.struct_libqt_string{
            .len = password.len,
            .data = @constCast(password.ptr),
        };
        qtc.QAuthenticator_SetPassword(@ptrCast(self), password_str);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qauthenticator.html#realm)
    ///
    /// ``` self: QtC.QAuthenticator, allocator: std.mem.Allocator ```
    pub fn Realm(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QAuthenticator_Realm(@ptrCast(self));
        defer qtc.libqt_string_free(@constCast(&_str));
        const _ret = allocator.alloc(u8, _str.len) catch @panic("qauthenticator.Realm: Memory allocation failed");
        for (0.._str.len) |_i| {
            _ret[_i] = _str.data[_i];
        }
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qauthenticator.html#setRealm)
    ///
    /// ``` self: QtC.QAuthenticator, realm: []const u8 ```
    pub fn SetRealm(self: ?*anyopaque, realm: []const u8) void {
        const realm_str = qtc.struct_libqt_string{
            .len = realm.len,
            .data = @constCast(realm.ptr),
        };
        qtc.QAuthenticator_SetRealm(@ptrCast(self), realm_str);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qauthenticator.html#option)
    ///
    /// ``` self: QtC.QAuthenticator, opt: []const u8 ```
    pub fn Option(self: ?*anyopaque, opt: []const u8) QtC.QVariant {
        const opt_str = qtc.struct_libqt_string{
            .len = opt.len,
            .data = @constCast(opt.ptr),
        };
        return qtc.QAuthenticator_Option(@ptrCast(self), opt_str);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qauthenticator.html#options)
    ///
    /// ``` self: QtC.QAuthenticator, allocator: std.mem.Allocator ```
    pub fn Options(self: ?*anyopaque, allocator: std.mem.Allocator) map_constu8_qtcqvariant {
        const _map: qtc.struct_libqt_map = qtc.QAuthenticator_Options(@ptrCast(self));
        var _ret: map_constu8_qtcqvariant = .empty;
        defer {
            const _keys: [*]qtc.struct_libqt_string = @ptrCast(@alignCast(_map.keys));
            for (0.._map.len) |_i| {
                qtc.libqt_free(_keys[_i].data);
            }
            qtc.libqt_free(_map.keys);
            qtc.libqt_free(_map.values);
        }
        const _keys: [*]qtc.struct_libqt_string = @ptrCast(@alignCast(_map.keys));
        const _values: [*]QtC.QVariant = @ptrCast(@alignCast(_map.values));
        var _i: usize = 0;
        while (_i < _map.len) : (_i += 1) {
            const _key = _keys[_i];
            const _entry_slice = std.mem.span(_key.data);
            const _value = _values[_i];
            _ret.put(allocator, _entry_slice, _value) catch @panic("qauthenticator.Options: Memory allocation failed");
        }
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qauthenticator.html#setOption)
    ///
    /// ``` self: QtC.QAuthenticator, opt: []const u8, value: QtC.QVariant ```
    pub fn SetOption(self: ?*anyopaque, opt: []const u8, value: ?*anyopaque) void {
        const opt_str = qtc.struct_libqt_string{
            .len = opt.len,
            .data = @constCast(opt.ptr),
        };
        qtc.QAuthenticator_SetOption(@ptrCast(self), opt_str, @ptrCast(value));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qauthenticator.html#isNull)
    ///
    /// ``` self: QtC.QAuthenticator ```
    pub fn IsNull(self: ?*anyopaque) bool {
        return qtc.QAuthenticator_IsNull(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qauthenticator.html#detach)
    ///
    /// ``` self: QtC.QAuthenticator ```
    pub fn Detach(self: ?*anyopaque) void {
        qtc.QAuthenticator_Detach(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qauthenticator.html#dtor.QAuthenticator)
    ///
    /// Delete this object from C++ memory.
    ///
    /// ``` self: QtC.QAuthenticator ```
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QAuthenticator_Delete(@ptrCast(self));
    }
};

File: src/network/libqauthenticator.cpp
#include <QAuthenticator>
#include <QMap>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QVariant>
#include <qauthenticator.h>
#include "libqauthenticator.h"
#include "libqauthenticator.hxx"

QAuthenticator* QAuthenticator_new() {
    return new QAuthenticator();
}

QAuthenticator* QAuthenticator_new2(const QAuthenticator* other) {
    return new QAuthenticator(*other);
}

void QAuthenticator_OperatorAssign(QAuthenticator* self, const QAuthenticator* other) {
    self->operator=(*other);
}

bool QAuthenticator_OperatorEqual(const QAuthenticator* self, const QAuthenticator* other) {
    return (*self == *other);
}

bool QAuthenticator_OperatorNotEqual(const QAuthenticator* self, const QAuthenticator* other) {
    return (*self != *other);
}

libqt_string QAuthenticator_User(const QAuthenticator* self) {
    QString _ret = self->user();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc((_str.len + 1) * sizeof(char)));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

void QAuthenticator_SetUser(QAuthenticator* self, const libqt_string user) {
    QString user_QString = QString::fromUtf8(user.data, user.len);
    self->setUser(user_QString);
}

libqt_string QAuthenticator_Password(const QAuthenticator* self) {
    QString _ret = self->password();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc((_str.len + 1) * sizeof(char)));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

void QAuthenticator_SetPassword(QAuthenticator* self, const libqt_string password) {
    QString password_QString = QString::fromUtf8(password.data, password.len);
    self->setPassword(password_QString);
}

libqt_string QAuthenticator_Realm(const QAuthenticator* self) {
    QString _ret = self->realm();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc((_str.len + 1) * sizeof(char)));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

void QAuthenticator_SetRealm(QAuthenticator* self, const libqt_string realm) {
    QString realm_QString = QString::fromUtf8(realm.data, realm.len);
    self->setRealm(realm_QString);
}

QVariant* QAuthenticator_Option(const QAuthenticator* self, const libqt_string opt) {
    QString opt_QString = QString::fromUtf8(opt.data, opt.len);
    return new QVariant(self->option(opt_QString));
}

libqt_map /* of libqt_string to QVariant* */ QAuthenticator_Options(const QAuthenticator* self) {
    QVariantHash _ret = self->options();
    // Convert QMap<> from C++ memory to manually-managed C memory
    libqt_string* _karr = static_cast<libqt_string*>(malloc(sizeof(libqt_string) * _ret.size()));
    QVariant** _varr = static_cast<QVariant**>(malloc(sizeof(QVariant*) * _ret.size()));
    int _ctr = 0;
    for (auto _itr = _ret.keyValueBegin(); _itr != _ret.keyValueEnd(); ++_itr) {
        QString _hashkey_ret = _itr->first;
        // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
        QByteArray _hashkey_b = _hashkey_ret.toUtf8();
        libqt_string _hashkey_str;
        _hashkey_str.len = _hashkey_b.length();
        _hashkey_str.data = static_cast<const char*>(malloc((_hashkey_str.len + 1) * sizeof(char)));
        memcpy((void*)_hashkey_str.data, _hashkey_b.data(), _hashkey_str.len);
        ((char*)_hashkey_str.data)[_hashkey_str.len] = '\0';
        _karr[_ctr] = _hashkey_str;
        _varr[_ctr] = new QVariant(_itr->second);
        _ctr++;
    }
    libqt_map _out;
    _out.len = _ret.size();
    _out.keys = static_cast<void*>(_karr);
    _out.values = static_cast<void*>(_varr);
    return _out;
}

void QAuthenticator_SetOption(QAuthenticator* self, const libqt_string opt, const QVariant* value) {
    QString opt_QString = QString::fromUtf8(opt.data, opt.len);
    self->setOption(opt_QString, *value);
}

bool QAuthenticator_IsNull(const QAuthenticator* self) {
    return self->isNull();
}

void QAuthenticator_Detach(QAuthenticator* self) {
    self->detach();
}

void QAuthenticator_Delete(QAuthenticator* self) {
    delete self;
}
